File: 1.21.9/src/main/java/com/bleudev/nine_lifes/util/RegistryUtils.java
package com.bleudev.nine_lifes.util;

import com.mojang.serialization.MapCodec;
import net.minecraft.entity.Entity;
import net.minecraft.entity.EntityType;
import net.minecraft.entity.effect.StatusEffect;
import net.minecraft.item.consume.ConsumeEffect;
import net.minecraft.network.RegistryByteBuf;
import net.minecraft.network.codec.PacketCodec;
import net.minecraft.registry.Registries;
import net.minecraft.registry.Registry;
import net.minecraft.registry.RegistryKey;
import net.minecraft.registry.RegistryKeys;
import net.minecraft.registry.entry.RegistryEntry;
import net.minecraft.util.Identifier;

import static com.bleudev.nine_lifes.NineLifesConst.MOD_ID;

public class RegistryUtils {
    public static Identifier get(String name) {
        return Identifier.of(MOD_ID, name);
    }

    public static RegistryEntry<StatusEffect> registerStatusEffect(String name, StatusEffect effect) {
        return RegistryEntry.of(Registry.register(Registries.STATUS_EFFECT, get(name), effect));
    }

    public static <T extends ConsumeEffect> ConsumeEffect.Type<T> registerConsumeEffectType(String name, MapCodec<T> codec, PacketCodec<RegistryByteBuf, T> packetCodec) {
        return Registry.register(Registries.CONSUME_EFFECT_TYPE, get(name), new ConsumeEffect.Type<>(codec, packetCodec));
    }

    public static <T extends Entity> EntityType<T> registerEntity(String path, EntityType.Builder<T> type) {
        RegistryKey<EntityType<?>> registryKey = RegistryKey.of(RegistryKeys.ENTITY_TYPE, Identifier.of(MOD_ID, path));
        return Registry.register(
            Registries.ENTITY_TYPE,
            Identifier.of(MOD_ID, path),
            type.build(registryKey)
        );
    }
}

File: 1.21.9/src/main/kotlin/com/bleudev/nine_lifes/util/RegistryUtils.kt
package com.bleudev.nine_lifes.util

import com.bleudev.nine_lifes.MOD_ID
import net.minecraft.core.Holder
import net.minecraft.core.Registry
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.core.registries.Registries
import net.minecraft.resources.ResourceKey
import net.minecraft.resources.ResourceLocation
import net.minecraft.world.effect.MobEffect
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.EntityType

fun createResourceLocation(path: String): ResourceLocation = ResourceLocation.fromNamespaceAndPath(MOD_ID, path)

fun registerMobEffect(name: String, effect: MobEffect): Holder<MobEffect> =
    Holder.direct(Registry.register(BuiltInRegistries.MOB_EFFECT, name, effect))

fun <T : Entity> registerEntity(path: String, type: EntityType.Builder<T>): EntityType<T> {
    val key = ResourceKey.create(Registries.ENTITY_TYPE, createResourceLocation(path))
    return Registry.register(BuiltInRegistries.ENTITY_TYPE, path, type.build(key))
}